homeassistant/components/history/__init__.py

```diff
@@ -293,7 +306,60 @@ def _states_to_json(
 
     # Append all changes to it
     for ent_id, group in groupby(states, lambda state: state.entity_id):
-        result[ent_id].extend(group)
+        domain = split_entity_id(ent_id)[0]
+        ent_results = result[ent_id]
+        if minimal_response and domain not in NEED_ATTRIBUTE_DOMAINS:
+            # With minimal response we only provide a native
+            # State for the first and last response. All the states
+            # in-between only provide the "state" and the
+            # "last_changed".
+            if not ent_results:
+                ent_results.append(next(group).to_native())
+
+            initial_state = ent_results[-1]
+            prev_state = ent_results[-1]
+            initial_state_count = len(ent_results)
+
+            for db_state in group:
+                if ATTR_HIDDEN in db_state.attributes and db_state.to_native().attributes.get(
+                    ATTR_HIDDEN, False
+                ):
+                    continue
+
+                # With minimal response we do not care about attribute
+                # changes so we can filter out duplicate states
+                if db_state.state == prev_state.state:
+                    continue
+
+                ent_results.append(
+                    {
+                        STATE_KEY: db_state.state,
+                        LAST_CHANGED_KEY: f"{str(db_state.last_changed).replace(' ','T').split('.')[0]}{DB_TIMEZONE}",
+                    }
+                )
+                prev_state = db_state
+
+            if (
+                prev_state
+                and prev_state != initial_state
+                and len(ent_results) != initial_state_count
+            ):
+                # There was at least one state change
+                # replace the last minimal state with
+                # a full state
+                ent_results[-1] = prev_state.to_native()
+        else:
+            ent_results.extend(
+                [
+                    native_state
+                    for native_state in (db_state.to_native() for db_state in group)
+                    if (
+                        domain != SCRIPT_DOMAIN
+                        or native_state.attributes.get(ATTR_CAN_CANCEL)
+                    )
+                    and not native_state.attributes.get(ATTR_HIDDEN, False)
+                ]
+            )
```

Comment:
Since this code is less complex, would it make sense to move this up in the for loop and add a `continue`, so the code for minimal results has less indenation

Reply:
Yes, very good idea, done!